documents/extension/funding/signatures.go

```diff
@@ -0,0 +1,182 @@
+
+	"github.com/centrifuge/go-centrifuge/config"
+	"github.com/centrifuge/go-centrifuge/contextutil"
+	"github.com/centrifuge/go-centrifuge/documents"
+	"github.com/centrifuge/go-centrifuge/errors"
+	"github.com/centrifuge/go-centrifuge/identity"
+	clientfundingpb "github.com/centrifuge/go-centrifuge/protobufs/gen/go/funding"
+	"github.com/centrifuge/go-centrifuge/utils"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+)
+
+func (s service) createSignAttrs(model documents.Model, idxFunding string, selfDID identity.DID, account config.Account) ([]documents.Attribute, error) {
+	var attributes []documents.Attribute
+	data, err := s.deriveFundingData(model, idxFunding)
+	if err != nil {
+		return nil, err
+	}
+
+	signMsg, err := json.Marshal(data)
+	if err != nil {
+		return nil, ErrJSON
+	}
+
+	// example "funding_agreement[2].signatures"
+	sLabel := generateLabel(fundingFieldKey, idxFunding, fundingSignatures)
+	attrIdx, err := incrementArrayAttrIDX(model, sLabel)
+	if err != nil {
+		return nil, err
+	}
+	attributes = append(attributes, attrIdx)
+
+	// example: "funding_agreement[2].signatures[4]"
+	sFieldLabel := generateLabel(generateLabel(fundingFieldKey, idxFunding, "")+fundingSignaturesFieldKey, attrIdx.Value.Int256.String(), "")
+
+	attrSign, err := documents.NewSignedAttribute(sFieldLabel, selfDID, account, model, signMsg)
+	if err != nil {
+		return nil, err
+	}
+	attributes = append(attributes, attrSign)
+
+	return attributes, nil
+}
+
+// Sign adds a signature to an existing document
+func (s service) Sign(ctx context.Context, fundingID string, identifier []byte) (documents.Model, error) {
+	selfDID, err := contextutil.AccountDID(ctx)
+	if err != nil {
+		return nil, errors.NewTypedError(documents.ErrDocumentConfigAccountID, err)
+	}
+
+	account, err := contextutil.Account(ctx)
+	if err != nil {
+		return nil, errors.NewTypedError(documents.ErrDocumentConfigAccountID, err)
+	}
+
+	model, err := s.Service.GetCurrentVersion(ctx, identifier)
+	if err != nil {
+		return nil, documents.ErrDocumentNotFound
+	}
+
+	idxFunding, err := s.findFundingIDX(model, fundingID)
+	if err != nil {
+		return nil, ErrFundingNotFound
+	}
+
+	attributes, err := s.createSignAttrs(model, idxFunding, selfDID, account)
+	if err != nil {
+		return nil, err
+	}
+
+	err = model.AddAttributes(attributes...)
+	if err != nil {
+		return nil, err
+	}
+
+	return model, nil
+}
+
+func (s service) validateValueOfSignAttr(funding *Data, signAttr documents.Attribute) (bool, error) {
+	value, err := json.Marshal(funding)
+	if err != nil {
+		return false, ErrJSON
+	}
+	return utils.IsSameByteSlice(value, signAttr.Value.Signed.Value), nil
+}
+
+func (s service) validateSignedFundingVersion(ctx context.Context, identifier []byte, fundingID string, signAttr documents.Attribute) (*clientfundingpb.FundingSignature, error) {
+	did := signAttr.Value.Signed.Identity
+	signedDocVersion, err := s.Service.GetVersion(ctx, identifier, signAttr.Value.Signed.DocumentVersion)
```

Comment:
I believe we are loading again the model, knowing that the parent invoker had it loaded?

Reply:
Resolved.